### PR TITLE
feat!: drop React 17 support

### PR DESCRIPTION
**Pull request by MichaelDeBoey**
@pcattori & @ryanflorence mentioned in #7021 that we would do this

### CODE REVIEW

File: docs/guides/typescript.md [hunk 1]
```diff
@@ -24,14 +24,14 @@ The Remix compiler will not do any type checking (it simply removes the types).
     "@remix-run/node": "latest",
     "@remix-run/react": "latest",
     "@remix-run/serve": "latest",
-    "react": "^17.0.2",
-    "react-dom": "^17.0.2"
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0"
```
Comment (brophdawg11) [inline docs/guides/typescript.md:28]: Is there a specific reason we don't use `^18.0.0` here?

**Comment by brophdawg11** [comment]
In order to support optional entry files and `npx remix reveal` - we had some logic in `@remix-run/dev` that detects the react version and uses/outputs react 17/18-compatible versions of entry.server/entry.client.  I removed all of that in [399f697](https://github.com/remix-run/remix/pull/7121/commits/399f6973e6030deb78b19b0d8eda5baed169b6d5) now that everything is react 18 + streaming.

File: packages/remix-dev/cli/commands.ts [hunk 2]
```diff
@@ -251,26 +236,11 @@ export async function generateEntry(
...
-  let clientRenderer = deps["@remix-run/react"] ? "react" : undefined;
-
-  if (!clientRenderer) {
-    console.error(
-      colors.error(
-        `Could not determine runtime. Please install the following: @remix-run/react`
-      )
-    );
-    return;
-  }
```
Comment (brophdawg11) [inline packages/remix-dev/cli/commands.ts:263]: This didn't seem necessary - we can assume React in Remix apps for now and the foreseeable future

File: packages/remix-dev/cli/commands.ts [hunk 3]
```diff
@@ -251,26 +236,11 @@ export async function generateEntry(
...
-  let defaultEntryClient = path.resolve(
-    defaultsDirectory,
-    `entry.client.${clientRenderer}-${type}.tsx`
-  );
+  let defaultEntryClient = path.resolve(defaultsDirectory, `entry.client.tsx`);
   let defaultEntryServer = path.resolve(
     defaultsDirectory,
-    serverRuntime,
-    `entry.server.${clientRenderer}-${type}.tsx`
+    `entry.server.${serverRuntime}.tsx`
```
Comment (brophdawg11) [inline packages/remix-dev/cli/commands.ts:243]: now that we only have one file per server runtime, I flattened them from:

`cloudflare/entry.server.react-stream.tsx -> entry.server.cloudflare.tsx`

**Review by pcattori — APPROVED** [review]

**Pull request merged**